# Machine Learning - Continuous Assessment - Class Group: TU060 <a class="tocSkip">

# MSc in Computer Science - Data Science (Part Time) <a class="tocSkip">

# Student Name: Ciaran Finnegan  <a class="tocSkip">

# Student No: D21124026 <a class="tocSkip">

#### Student EMail: D21124026@mytudublin.ie <a class="tocSkip">

## May 2022<a class="tocSkip">

# 

# H&M Personalised Fashion Recommendations <a class="tocSkip">


#### This project takes the datasets provided by Kaggle for the H&M Personalised Fashion Recommendations competition.  <a class="tocSkip">

Ipso forum...

## <a class="tocSkip">

#### The CA follows a n-part structure. <a class="tocSkip">

Section 1: Ipso forum...

Section 2: Ipso forum...

Section 3: Ipso forum...


### <a class="tocSkip">

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Jupyter-Notebook-Set-Up" data-toc-modified-id="Jupyter-Notebook-Set-Up-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Jupyter Notebook Set Up</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Import Libraries</a></span><ul class="toc-item"><li><span><a href="#General-Utility-Libraries" data-toc-modified-id="General-Utility-Libraries-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>General Utility Libraries</a></span></li><li><span><a href="#Panda-+-Numpy" data-toc-modified-id="Panda-+-Numpy-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Panda + Numpy</a></span></li><li><span><a href="#Matplotlib-Library" data-toc-modified-id="Matplotlib-Library-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Matplotlib Library</a></span></li><li><span><a href="#Plotly-Libraries" data-toc-modified-id="Plotly-Libraries-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Plotly Libraries</a></span></li></ul></li></ul></li><li><span><a href="#Project-Data-Import,-Analysis-and-Preparation" data-toc-modified-id="Project-Data-Import,-Analysis-and-Preparation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Project Data Import, Analysis and Preparation</a></span><ul class="toc-item"><li><span><a href="#Load-and-Sample-Kaggle-H&amp;M-Datasets" data-toc-modified-id="Load-and-Sample-Kaggle-H&amp;M-Datasets-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load and Sample Kaggle H&amp;M Datasets</a></span><ul class="toc-item"><li><span><a href="#Set-Up-File-Paths" data-toc-modified-id="Set-Up-File-Paths-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Set Up File Paths</a></span></li><li><span><a href="#H&amp;M-Transaction-Dataset" data-toc-modified-id="H&amp;M-Transaction-Dataset-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>H&amp;M Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Articles-Dataset" data-toc-modified-id="H&amp;M-Articles-Dataset-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>H&amp;M Articles Dataset</a></span></li><li><span><a href="#H&amp;M-Customers-Dataset" data-toc-modified-id="H&amp;M-Customers-Dataset-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>H&amp;M Customers Dataset</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#H&amp;M-Sampled-Transaction-Dataset" data-toc-modified-id="H&amp;M-Sampled-Transaction-Dataset-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>H&amp;M Sampled Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Customer-Dataset" data-toc-modified-id="H&amp;M-Sampled-Customer-Dataset-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>H&amp;M Sampled Customer Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Articles-Dataset" data-toc-modified-id="H&amp;M-Sampled-Articles-Dataset-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>H&amp;M Sampled Articles Dataset</a></span></li></ul></li><li><span><a href="#Data-Preparation-and-Feature-Enrichment" data-toc-modified-id="Data-Preparation-and-Feature-Enrichment-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Data Preparation and Feature Enrichment</a></span><ul class="toc-item"><li><span><a href="#Transaction-Data-Feature-Enrichment-(Sample)" data-toc-modified-id="Transaction-Data-Feature-Enrichment-(Sample)-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Transaction Data Feature Enrichment (Sample)</a></span></li><li><span><a href="#Extend-Dataframe-with-Customer-Data" data-toc-modified-id="Extend-Dataframe-with-Customer-Data-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Extend Dataframe with Customer Data</a></span></li><li><span><a href="#Extend-Dataframe-with-Articles-Data" data-toc-modified-id="Extend-Dataframe-with-Articles-Data-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Extend Dataframe with Articles Data</a></span></li><li><span><a href="#Final-Data-Preparation-Steps" data-toc-modified-id="Final-Data-Preparation-Steps-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Final Data Preparation Steps</a></span></li></ul></li></ul></li><li><span><a href="#Split-Data:-Train-and-Test-Sets" data-toc-modified-id="Split-Data:-Train-and-Test-Sets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Split Data: Train and Test Sets</a></span></li><li><span><a href="#Model-Preparation" data-toc-modified-id="Model-Preparation-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model Preparation</a></span></li><li><span><a href="#K-Nearest-Neighbor-Recommendations" data-toc-modified-id="K-Nearest-Neighbor-Recommendations-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>K Nearest Neighbor Recommendations</a></span></li><li><span><a href="#Submission" data-toc-modified-id="Submission-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Submission</a></span></li><li><span><a href="#References" data-toc-modified-id="References-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Jupyter Notebook Set Up

### Import Libraries

Note: It is assumed that the matplotlib and plotly library has been installed on the user's machine before it is available for import.

```bash
> pip install matplotlib
> pip install plotly
```


#### General Utility Libraries

In [1]:
import os

#### Panda + Numpy

In [2]:
# Python libraries for data manipulation
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import NearestNeighbors

Set up additional libraries for graphical presentations. A number of earlier graphs use Matplotlib. In later sections, Plotly graphs are used.

#### Matplotlib Library

In [3]:
#import matplotlib.pyplot as plt

#### Plotly Libraries

In [4]:
# For interactive graphics and global displays
#import plotly.express as px
#import plotly.graph_objects as go

### <a class="tocSkip">

In [5]:
import warnings
warnings.filterwarnings("ignore")

### <a class="tocSkip">

## Project Data Import, Analysis and Preparation

### Load and Sample Kaggle H&M Datasets

### The assignment starts with a Kaggle H&M Personalised Fashion Recommendations datasets. <a class="tocSkip">

#### Set Up File Paths 

In [6]:
# Set up a Global path file variable to use when loading each dataset
Path_To_Inputs = '../input'

#### H&M Transaction Dataset 

In [7]:
# Load H&M Transaction dataset
df_trxns = pd.read_csv(os.path.join(Path_To_Inputs,'transactions_train.csv'))

In [8]:
# Display Number of Rows, Number of Columns in Transaction dataset prior to Sampling
df_trxns.shape

(31788324, 5)

In [9]:
# Extract random sample of Transactions (without replacement) - 0.1% of records
df_trxns_sample = df_trxns.sample(frac=0.001, replace=False)
## NEW CODE - Check this out
## df_trxns_sample.set_index(inplace = True)

In [10]:
# Display Number of Rows, Number of Columns in Transaction dataset after Sampling
df_trxns_sample.shape

(31788, 5)

In [11]:
# Check out the index
df_trxns_sample.head()

t_dat                                        customer_id  \
18606403  2019-11-03  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   
26192852  2020-05-23  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   
11014491  2019-05-29  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   
24927220  2020-04-22  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   
21126624  2020-01-10  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   

          article_id     price  sales_channel_id  
18606403   820563003  0.027102                 2  
26192852   854883003  0.030492                 2  
11014491   683001006  0.015237                 1  
24927220   867044001  0.033881                 2  
21126624   837014001  0.022017                 2

#### H&M Articles Dataset 

In [12]:
# Read file, located in 'input' folder above this notebook, for articles data - containing meta data on purchasable items.
df_Articles = pd.read_csv(os.path.join(Path_To_Inputs,'articles.csv'))

In [13]:
# Display Number of Rows, Number of Columns in Articles dataset
df_Articles.shape

(105542, 25)

In [14]:
# Display basic file data for the initial Kaggle H&M 'articles' datatset
# df_Articles.head()

In [15]:
# Extract only the H&M articles contained in the sample Transaction dataset 
articleIDs_in_Trxn_Sample  = set(df_trxns_sample["article_id"])

In [16]:
#Count Article Ids in Transaction Sample
#len(articleIDs_in_Trxn_Sample)

In [17]:
# Create dataset that only contains H&M articles in the sample Transaction dataset
df_articles_sample = df_Articles[df_Articles["article_id"].isin(articleIDs_in_Trxn_Sample)]

In [18]:
# Display Number of Rows, Number of Columns in Sample Articles dataset
df_articles_sample.shape

(18502, 25)

#### H&M Customers Dataset 

In [19]:
# Load H&M Customer dataset
df_customers = pd.read_csv(os.path.join(Path_To_Inputs,'customers.csv'))

In [20]:
# Display Number of Rows, Number of Columns in Customers dataset prior to Sampling
df_customers.shape

(1371980, 7)

In [21]:
# Extract only the H&M Customers contained in the sample Transaction dataset 
custIDs_in_Trxn_Sample  = set(df_trxns_sample["customer_id"])

In [22]:
#Count Customer Ids in Transaction Sample
len(custIDs_in_Trxn_Sample)

30451

In [23]:
# Create dataset that only contains H&M Customers from the sample Transaction dataset
df_customers_sample = df_customers[df_customers["customer_id"].isin(custIDs_in_Trxn_Sample)]

In [24]:
# Display Number of Rows, Number of Columns in Sample Customers dataset
df_customers_sample.shape

(30451, 7)

###### 

### Exploratory Data Analysis

### Display key data attributes and distributions in Sampled H&M datasets. <a class="tocSkip">

#### H&M Sampled Transaction Dataset 

In [25]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - after Sampling of original dataset
df_trxns_sample.head()

t_dat                                        customer_id  \
18606403  2019-11-03  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   
26192852  2020-05-23  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   
11014491  2019-05-29  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   
24927220  2020-04-22  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   
21126624  2020-01-10  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   

          article_id     price  sales_channel_id  
18606403   820563003  0.027102                 2  
26192852   854883003  0.030492                 2  
11014491   683001006  0.015237                 1  
24927220   867044001  0.033881                 2  
21126624   837014001  0.022017                 2

In [26]:
# Display datatype of colums in original H&M Transaction dataset
df_trxns_sample.dtypes

t_dat                object
customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
dtype: object

#### H&M Sampled Customer Dataset 

In [27]:
# Display basic column information for the initial Kaggle 'customers' datatset - after Sampling of original dataset
df_customers_sample.head()

customer_id   FN  Active  \
13   00009d946eec3ea54add5ba56d5210ea898def4b46c685...  1.0     1.0   
22   0000f2ea26b7f0a9175f428c8cf7743e9e10e193465ecd...  NaN     NaN   
128  00058592fc65afabbb00b1bb7d33c6b221d00c6a98c621...  NaN     NaN   
203  0008968c0d451dbc5a9968da03196fe20051965edde741...  1.0     1.0   
304  000e2691ade6addc974a2d8767def5b5272d1c62883cc0...  NaN     NaN   

    club_member_status fashion_news_frequency   age  \
13              ACTIVE              Regularly  56.0   
22              ACTIVE                   NONE  48.0   
128             ACTIVE                   NONE  24.0   
203             ACTIVE              Regularly  50.0   
304             ACTIVE                   NONE  28.0   

                                           postal_code  
13   b31984b20a8c478de38eaf113c581ff64e63c4242e607b...  
22   11404208b30249e22baed5c16782671d46b5e5e0690ab7...  
128  09c5be6b8bd6e0c04be46c5062a703db15d9d0d6429291...  
203  860a370c01b1eeebe3f3c78241f3236c79b0c41669c19e...  
304  3bb3547ebd518e6d2741f19f724bed8da1b069dfa66224...

In [28]:
df_customers_sample.dtypes

customer_id                object
FN                        float64
Active                    float64
club_member_status         object
fashion_news_frequency     object
age                       float64
postal_code                object
dtype: object

#### H&M Sampled Articles Dataset 

In [29]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code                  prod_name  product_type_no  \
0   108775015        108775                  Strap top              253   
1   108775044        108775                  Strap top              253   
5   110065011        110065          OP T-shirt (Idro)              306   
6   111565001        111565        20 den 1p Stockings              304   
8   111586001        111586  Shape Up 30 den 1p Tights              273   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
5               Bra           Underwear                  1010016   
6  Underwear Tights      Socks & Tights                  1010016   
8   Leggings/Tights  Garment Lower body                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
5                     Solid                 12       Light Beige  ...   
6                     Solid                  9             Black  ...   
8                     Solid                  9             Black  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
5   Clean Lingerie          B  Lingeries/Tights              1   
6     Tights basic          B  Lingeries/Tights              1   
8     Tights basic          B  Lingeries/Tights              1   

   index_group_name section_no                    section_name  \
0        Ladieswear         16          Womens Everyday Basics   
1        Ladieswear         16          Womens Everyday Basics   
5        Ladieswear         61                 Womens Lingerie   
6        Ladieswear         62  Womens Nightwear, Socks & Tigh   
8        Ladieswear         62  Womens Nightwear, Socks & Tigh   

  garment_group_no  garment_group_name  \
0             1002        Jersey Basic   
1             1002        Jersey Basic   
5             1017   Under-, Nightwear   
6             1021    Socks and Tights   
8             1021    Socks and Tights   

                                         detail_desc  
0            Jersey top with narrow shoulder straps.  
1            Jersey top with narrow shoulder straps.  
5  Microfibre T-shirt bra with underwired, moulde...  
6  Semi shiny nylon stockings with a wide, reinfo...  
8  Tights with built-in support to lift the botto...  

[5 rows x 25 columns]

###### 

### Data Preparation and Feature Enrichment

### Construct a dataset upon which to build ML models. <a class="tocSkip">

### Begin with Transaction Dataset (sampled) and merge in customer and article data. <a class="tocSkip">

#### Transaction Data Feature Enrichment (Sample)

In [30]:
# Create new dataset for modelling. Initial basis for new dataset is the sampled transaction dataset.
df_HM_CustTrxn = df_trxns_sample

In [31]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
18606403  2019-11-03  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   
26192852  2020-05-23  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   
11014491  2019-05-29  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   
24927220  2020-04-22  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   
21126624  2020-01-10  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   

          article_id     price  sales_channel_id  
18606403   820563003  0.027102                 2  
26192852   854883003  0.030492                 2  
11014491   683001006  0.015237                 1  
24927220   867044001  0.033881                 2  
21126624   837014001  0.022017                 2

In [32]:
# Decompose Datetime information 
df_HM_CustTrxn['t_dat'] = pd.to_datetime(df_HM_CustTrxn['t_dat'], format="%Y-%m-%d")

# Add Column for Transaction Year and convert to integer value
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['t_dat'].dt.strftime('%Y')
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['txn_year'].astype(str).astype(int)

# Add Column for Transaction Month and convert to integer value
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['t_dat'].dt.strftime('%m')
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['txn_mth'].astype(str).astype(int)

# Add Column for Transaction Day and convert to integer value
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['t_dat'].dt.strftime('%d')
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['txn_day'].astype(str).astype(int)

In [33]:
# Filter out column to drop timestamp attribute
df_HM_CustTrxn = df_HM_CustTrxn.iloc[:,1:8]

In [34]:
df_HM_CustTrxn.head()

customer_id  article_id  \
18606403  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003   
26192852  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003   
11014491  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006   
24927220  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001   
21126624  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001   

             price  sales_channel_id  txn_year  txn_mth  txn_day  
18606403  0.027102                 2      2019       11        3  
26192852  0.030492                 2      2020        5       23  
11014491  0.015237                 1      2019        5       29  
24927220  0.033881                 2      2020        4       22  
21126624  0.022017                 2      2020        1       10

In [35]:
df_HM_CustTrxn.dtypes

customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
txn_year              int32
txn_mth               int32
txn_day               int32
dtype: object

In [36]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the Trxn (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the Trxn (sample) dataframe? 
 customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
dtype: bool


[None]

#### Extend Dataframe with Customer Data

In [37]:
# Confirm dimensions of Trxn (sample) dataframe BEFORE merging with Customer data
df_HM_CustTrxn.shape

(31788, 7)

#### Merge Customer Attributes into the Trxn (Sample) Dataframe. <a class="tocSkip">

In [38]:
# Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
df_HM_CustTrxn = pd.merge(df_HM_CustTrxn, df_customers_sample, how="left", on=["customer_id"])

In [39]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Customer data
df_HM_CustTrxn.shape

(31788, 13)

In [40]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day   FN  Active  \
0                 2      2019       11        3  1.0     1.0   
1                 2      2020        5       23  NaN     NaN   
2                 1      2019        5       29  NaN     NaN   
3                 2      2020        4       22  NaN     NaN   
4                 2      2020        1       10  NaN     NaN   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE              Regularly  27.0   
1             ACTIVE                   NONE  30.0   
2             ACTIVE                   NONE  28.0   
3             ACTIVE                   NONE  49.0   
4             ACTIVE                   NONE  41.0   

                                         postal_code  
0  0bfdb52d2de687265dc42f091ccd6cf324c75757273732...  
1  c47eb38049a361a1fb1f21aa1bb5d9f6c85e073862a632...  
2  2c29ae653a9282cce4151bd87643c907644e09541abc28...  
3  d5c7421f809240c05717ed726110b5c7a7d50fd55b6fb5...  
4  7eb4077cc65d97bd9196d6618d0cae5fc4f6c00005b572...

In [41]:
# Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
# There are very many missing values in the other Customer dataset columns, so these will be ignored
df_HM_CustTrxn = df_HM_CustTrxn.iloc[:, [0, 1, 2, 3, 4, 5, 6, 9, 10, 11, 12]]

In [42]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019       11        3             ACTIVE   
1                 2      2020        5       23             ACTIVE   
2                 1      2019        5       29             ACTIVE   
3                 2      2020        4       22             ACTIVE   
4                 2      2020        1       10             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  27.0   
1                   NONE  30.0   
2                   NONE  28.0   
3                   NONE  49.0   
4                   NONE  41.0   

                                         postal_code  
0  0bfdb52d2de687265dc42f091ccd6cf324c75757273732...  
1  c47eb38049a361a1fb1f21aa1bb5d9f6c85e073862a632...  
2  2c29ae653a9282cce4151bd87643c907644e09541abc28...  
3  d5c7421f809240c05717ed726110b5c7a7d50fd55b6fb5...  
4  7eb4077cc65d97bd9196d6618d0cae5fc4f6c00005b572...

In [43]:
df_HM_CustTrxn.shape

(31788, 11)

In [44]:
# Check for null values in the merged Trxn/Customer dataframe
[print('Any Null Values in the merged Trxn/Customer dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer dataframe? 
 customer_id               False
article_id                False
price                     False
sales_channel_id          False
txn_year                  False
txn_mth                   False
txn_day                   False
club_member_status         True
fashion_news_frequency     True
age                        True
postal_code               False
dtype: bool


[None]

#### Remove Rows with Missing Club Member Status Attributes. <a class="tocSkip">

In [45]:
# Check missing values in the club_member_status column  
df_null_club_member_status = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].isnull())]

[print('Numbers of rows with missing club member status?','\n' ,len(df_null_club_member_status.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing club member status attributes?\n{:06.3f}%'
                               .format((len(df_null_club_member_status.index)/len(df_HM_CustTrxn.index))*100))]


# For simplicity the rows with missing values in the club member status will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the club member status
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the club member status column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Numbers of rows with missing club member status? 
 63

Precentage of total rows in Trxn dataframe with missing club member status attributes?
00.198%

Current Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed 
 (31725, 11)


[None]

#### Remove Rows with Missing Fashion News Frequency Attributes. <a class="tocSkip">

In [46]:
# Check missing values in the club_member_status column  
df_null_fn_freq = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].isnull())]

[print('Numbers of rows with missing fashion_news_frequency values?','\n' ,len(df_null_fn_freq.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?\n{:06.3f}%'
                               .format((len(df_null_fn_freq.index)/len(df_HM_CustTrxn.index))*100))]

# For simplicity the rows with missing values in the fashion_news_frequency column will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the fashion_news_frequency column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the fashion_news_frequency column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]

Numbers of rows with missing fashion_news_frequency values? 
 149

Precentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?
00.470%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed 
 (31576, 11)


[None]

#### Remove Rows with Missing Age Attributes. <a class="tocSkip">

In [47]:
# Check missing values in the age column  
df_null_age = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].isnull())]

[print('Numbers of rows with missing age values?','\n' ,len(df_null_age.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing age attributes?\n{:06.3f}%'
                               .format((len(df_null_age.index)/len(df_HM_CustTrxn.index))*100))]

# Re-Build Txn/Customer Dataframe with no missing values in the age column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].notnull())]


# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the age column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed','\n' ,df_HM_CustTrxn.shape)]

Numbers of rows with missing age values? 
 138

Precentage of total rows in Trxn dataframe with missing age attributes?
00.437%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed 
 (31438, 11)


[None]

#### Encode the club_member_status + fashion_news_frequency attributes <a class="tocSkip">

In [48]:
# The categorical values for club_member_status must be encoded to numberical values
df_encoded_member_stats = pd.get_dummies(df_HM_CustTrxn['club_member_status'])

In [49]:
df_encoded_member_stats

ACTIVE  LEFT CLUB  PRE-CREATE
0           1          0           0
1           1          0           0
2           1          0           0
3           1          0           0
4           1          0           0
...       ...        ...         ...
31782       1          0           0
31783       1          0           0
31784       1          0           0
31785       1          0           0
31787       1          0           0

[31438 rows x 3 columns]

In [50]:
# Add Encoded club_member_status to our merged H&M dataframe
df_HM_CustTrxn = df_HM_CustTrxn.join(df_encoded_member_stats)

In [51]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019       11        3             ACTIVE   
1                 2      2020        5       23             ACTIVE   
2                 1      2019        5       29             ACTIVE   
3                 2      2020        4       22             ACTIVE   
4                 2      2020        1       10             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  27.0   
1                   NONE  30.0   
2                   NONE  28.0   
3                   NONE  49.0   
4                   NONE  41.0   

                                         postal_code  ACTIVE  LEFT CLUB  \
0  0bfdb52d2de687265dc42f091ccd6cf324c75757273732...       1          0   
1  c47eb38049a361a1fb1f21aa1bb5d9f6c85e073862a632...       1          0   
2  2c29ae653a9282cce4151bd87643c907644e09541abc28...       1          0   
3  d5c7421f809240c05717ed726110b5c7a7d50fd55b6fb5...       1          0   
4  7eb4077cc65d97bd9196d6618d0cae5fc4f6c00005b572...       1          0   

   PRE-CREATE  
0           0  
1           0  
2           0  
3           0  
4           0

In [52]:
# The categorical values for fashion_news_frequency must be encoded to numberical values
df_encoded_fn_freq = pd.get_dummies(df_HM_CustTrxn['fashion_news_frequency'])

In [53]:
df_encoded_fn_freq

Monthly  NONE  Regularly
0            0     0          1
1            0     1          0
2            0     1          0
3            0     1          0
4            0     1          0
...        ...   ...        ...
31782        0     0          1
31783        0     1          0
31784        0     1          0
31785        0     0          1
31787        0     1          0

[31438 rows x 3 columns]

In [54]:
# Add Encoded fashion_news_frequency to our merged H&M dataframe
df_HM_CustTrxn = df_HM_CustTrxn.join(df_encoded_fn_freq)

In [55]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019       11        3             ACTIVE   
1                 2      2020        5       23             ACTIVE   
2                 1      2019        5       29             ACTIVE   
3                 2      2020        4       22             ACTIVE   
4                 2      2020        1       10             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  27.0   
1                   NONE  30.0   
2                   NONE  28.0   
3                   NONE  49.0   
4                   NONE  41.0   

                                         postal_code  ACTIVE  LEFT CLUB  \
0  0bfdb52d2de687265dc42f091ccd6cf324c75757273732...       1          0   
1  c47eb38049a361a1fb1f21aa1bb5d9f6c85e073862a632...       1          0   
2  2c29ae653a9282cce4151bd87643c907644e09541abc28...       1          0   
3  d5c7421f809240c05717ed726110b5c7a7d50fd55b6fb5...       1          0   
4  7eb4077cc65d97bd9196d6618d0cae5fc4f6c00005b572...       1          0   

   PRE-CREATE  Monthly  NONE  Regularly  
0           0        0     0          1  
1           0        0     1          0  
2           0        0     1          0  
3           0        0     1          0  
4           0        0     1          0

In [56]:
# Remove the prior Categorical columns from the Trxn/Customer dataframe
# The 'inplace' parameter confirms that the drop operation takes place in the same dataframe (a new one is not created)
df_HM_CustTrxn.drop(["club_member_status","fashion_news_frequency"], axis=1, inplace=True)

In [57]:
# Drop Postal Code - a non numerical value that is not easily converted
df_HM_CustTrxn.drop(["postal_code"], axis=1, inplace=True)

In [58]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  LEFT CLUB  \
0                 2      2019       11        3  27.0       1          0   
1                 2      2020        5       23  30.0       1          0   
2                 1      2019        5       29  28.0       1          0   
3                 2      2020        4       22  49.0       1          0   
4                 2      2020        1       10  41.0       1          0   

   PRE-CREATE  Monthly  NONE  Regularly  
0           0        0     0          1  
1           0        0     1          0  
2           0        0     1          0  
3           0        0     1          0  
4           0        0     1          0

In [59]:
df_HM_CustTrxn.dtypes

customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
txn_year              int32
txn_mth               int32
txn_day               int32
age                 float64
ACTIVE                uint8
LEFT CLUB             uint8
PRE-CREATE            uint8
Monthly               uint8
NONE                  uint8
Regularly             uint8
dtype: object

In [60]:
# Confirm dimensions of Trxn (sample) dataframe AFTER encoding Categorical Values
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Current Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed 
 (31438, 14)


[None]

In [61]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the merged Trxn/Customer (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer (sample) dataframe? 
 customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
age                 False
ACTIVE              False
LEFT CLUB           False
PRE-CREATE          False
Monthly             False
NONE                False
Regularly           False
dtype: bool


[None]

#### Extend Dataframe with Articles Data

#### Remove 'Redundant' Categorical Columns in Articles (sample) Dataframe. <a class="tocSkip">

In [62]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code                  prod_name  product_type_no  \
0   108775015        108775                  Strap top              253   
1   108775044        108775                  Strap top              253   
5   110065011        110065          OP T-shirt (Idro)              306   
6   111565001        111565        20 den 1p Stockings              304   
8   111586001        111586  Shape Up 30 den 1p Tights              273   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
5               Bra           Underwear                  1010016   
6  Underwear Tights      Socks & Tights                  1010016   
8   Leggings/Tights  Garment Lower body                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
5                     Solid                 12       Light Beige  ...   
6                     Solid                  9             Black  ...   
8                     Solid                  9             Black  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
5   Clean Lingerie          B  Lingeries/Tights              1   
6     Tights basic          B  Lingeries/Tights              1   
8     Tights basic          B  Lingeries/Tights              1   

   index_group_name section_no                    section_name  \
0        Ladieswear         16          Womens Everyday Basics   
1        Ladieswear         16          Womens Everyday Basics   
5        Ladieswear         61                 Womens Lingerie   
6        Ladieswear         62  Womens Nightwear, Socks & Tigh   
8        Ladieswear         62  Womens Nightwear, Socks & Tigh   

  garment_group_no  garment_group_name  \
0             1002        Jersey Basic   
1             1002        Jersey Basic   
5             1017   Under-, Nightwear   
6             1021    Socks and Tights   
8             1021    Socks and Tights   

                                         detail_desc  
0            Jersey top with narrow shoulder straps.  
1            Jersey top with narrow shoulder straps.  
5  Microfibre T-shirt bra with underwired, moulde...  
6  Semi shiny nylon stockings with a wide, reinfo...  
8  Tights with built-in support to lift the botto...  

[5 rows x 25 columns]

In [63]:
# Remove 'redundant' categotical columns from Articles (sample) dataframe
# Consider a categorical column Redundant when is 'partnered' with a numerical code 
df_articles_sample.drop(["prod_name", 
                         "product_type_name", 
                         "graphical_appearance_name",
                         "colour_group_name",
                         "perceived_colour_value_name",
                         "perceived_colour_master_name",
                         "department_name",
                         "index_name",
                         "index_group_name",
                         "section_name",
                         "garment_group_name",
                         "detail_desc"], axis=1, inplace=True)

In [64]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
5   110065011        110065              306           Underwear   
6   111565001        111565              304      Socks & Tights   
8   111586001        111586              273  Garment Lower body   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
5                  1010016                 12                          1   
6                  1010016                  9                          4   
8                  1010016                  9                          4   

   perceived_colour_master_id  department_no index_code  index_group_no  \
0                           5           1676          A               1   
1                           9           1676          A               1   
5                          11           1339          B               1   
6                           5           3608          B               1   
8                           5           3608          B               1   

   section_no  garment_group_no  
0          16              1002  
1          16              1002  
5          61              1017  
6          62              1021  
8          62              1021

#### Use One-Hot-Encoding on remaining categorical columns. <a class="tocSkip">

In [65]:
# The categorical values for product_group_name + index_code must be encoded to numberical values
df_encoded_prd_grp_nm = pd.get_dummies(df_articles_sample['product_group_name'])
df_encoded_idx_cd     = pd.get_dummies(df_articles_sample['index_code'])

In [66]:
df_encoded_prd_grp_nm

Accessories  Bags  Cosmetic  Furniture  Garment Full body  \
0                 0     0         0          0                  0   
1                 0     0         0          0                  0   
5                 0     0         0          0                  0   
6                 0     0         0          0                  0   
8                 0     0         0          0                  0   
...             ...   ...       ...        ...                ...   
105398            0     0         0          0                  1   
105471            0     0         0          0                  0   
105485            0     0         0          0                  0   
105491            0     0         0          0                  0   
105505            0     0         0          0                  0   

        Garment Lower body  Garment Upper body  Items  Nightwear  Shoes  \
0                        0                   1      0          0      0   
1                        0                   1      0          0      0   
5                        0                   0      0          0      0   
6                        0                   0      0          0      0   
8                        1                   0      0          0      0   
...                    ...                 ...    ...        ...    ...   
105398                   0                   0      0          0      0   
105471                   1                   0      0          0      0   
105485                   0                   0      0          0      0   
105491                   0                   1      0          0      0   
105505                   0                   1      0          0      0   

        Socks & Tights  Swimwear  Underwear  Unknown  
0                    0         0          0        0  
1                    0         0          0        0  
5                    0         0          1        0  
6                    1         0          0        0  
8                    0         0          0        0  
...                ...       ...        ...      ...  
105398               0         0          0        0  
105471               0         0          0        0  
105485               0         0          1        0  
105491               0         0          0        0  
105505               0         0          0        0  

[18502 rows x 14 columns]

In [67]:
df_encoded_idx_cd

A  B  C  D  F  G  H  I  J  S
0       1  0  0  0  0  0  0  0  0  0
1       1  0  0  0  0  0  0  0  0  0
5       0  1  0  0  0  0  0  0  0  0
6       0  1  0  0  0  0  0  0  0  0
8       0  1  0  0  0  0  0  0  0  0
...    .. .. .. .. .. .. .. .. .. ..
105398  0  0  0  0  0  0  0  1  0  0
105471  1  0  0  0  0  0  0  0  0  0
105485  1  0  0  0  0  0  0  0  0  0
105491  0  0  0  1  0  0  0  0  0  0
105505  1  0  0  0  0  0  0  0  0  0

[18502 rows x 10 columns]

In [68]:
# Add Encoded product_group_name to our articles dataframe (sample)
df_articles_sample = df_articles_sample.join(df_encoded_prd_grp_nm)
df_articles_sample = df_articles_sample.join(df_encoded_idx_cd)

In [69]:
df_articles_sample.shape

(18502, 37)

In [70]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
5   110065011        110065              306           Underwear   
6   111565001        111565              304      Socks & Tights   
8   111586001        111586              273  Garment Lower body   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
5                  1010016                 12                          1   
6                  1010016                  9                          4   
8                  1010016                  9                          4   

   perceived_colour_master_id  department_no index_code  ...  A  B  C  D  F  \
0                           5           1676          A  ...  1  0  0  0  0   
1                           9           1676          A  ...  1  0  0  0  0   
5                          11           1339          B  ...  0  1  0  0  0   
6                           5           3608          B  ...  0  1  0  0  0   
8                           5           3608          B  ...  0  1  0  0  0   

   G  H  I  J  S  
0  0  0  0  0  0  
1  0  0  0  0  0  
5  0  0  0  0  0  
6  0  0  0  0  0  
8  0  0  0  0  0  

[5 rows x 37 columns]

In [71]:
# Remove the prior Categorical columns from the Articles dataframe
df_articles_sample.drop(["product_group_name","index_code"], axis=1, inplace=True)

In [72]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  graphical_appearance_no  \
0   108775015        108775              253                  1010016   
1   108775044        108775              253                  1010016   
5   110065011        110065              306                  1010016   
6   111565001        111565              304                  1010016   
8   111586001        111586              273                  1010016   

   colour_group_code  perceived_colour_value_id  perceived_colour_master_id  \
0                  9                          4                           5   
1                 10                          3                           9   
5                 12                          1                          11   
6                  9                          4                           5   
8                  9                          4                           5   

   department_no  index_group_no  section_no  ...  A  B  C  D  F  G  H  I  J  \
0           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
1           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
5           1339               1          61  ...  0  1  0  0  0  0  0  0  0   
6           3608               1          62  ...  0  1  0  0  0  0  0  0  0   
8           3608               1          62  ...  0  1  0  0  0  0  0  0  0   

   S  
0  0  
1  0  
5  0  
6  0  
8  0  

[5 rows x 35 columns]

In [73]:
df_articles_sample.dtypes

article_id                    int64
product_code                  int64
product_type_no               int64
graphical_appearance_no       int64
colour_group_code             int64
perceived_colour_value_id     int64
perceived_colour_master_id    int64
department_no                 int64
index_group_no                int64
section_no                    int64
garment_group_no              int64
Accessories                   uint8
Bags                          uint8
Cosmetic                      uint8
Furniture                     uint8
Garment Full body             uint8
Garment Lower body            uint8
Garment Upper body            uint8
Items                         uint8
Nightwear                     uint8
Shoes                         uint8
Socks & Tights                uint8
Swimwear                      uint8
Underwear                     uint8
Unknown                       uint8
A                             uint8
B                             uint8
C                           

#### Merge Trxn/Customers Dataframe with Articles Dataframe (sample). <a class="tocSkip">

In [74]:
# Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
df_HM_CustTrxnArtcls = pd.merge(df_HM_CustTrxn, df_articles_sample, how="left", on=["article_id"])

In [75]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Articles data
df_HM_CustTrxnArtcls.shape

(31438, 48)

In [76]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - AFTER merging with Articles data
df_HM_CustTrxnArtcls.head()

customer_id  article_id     price  \
0  8e2b74eb1c5523f185dea917b5583b77287adb2dcea69a...   820563003  0.027102   
1  b1fdcb46568ce4b8df5ce4ca6033aacf038a22e73d6567...   854883003  0.030492   
2  b8ec3298c681ee156de41e7b3ae853c38074d843d75af9...   683001006  0.015237   
3  a64d4624813554a34307d92cbced37d24dcefc6b060d94...   867044001  0.033881   
4  091a8a935b920bc68d79cbccbf17a14c8bb09a8f13d968...   837014001  0.022017   

   sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  LEFT CLUB  ...  \
0                 2      2019       11        3  27.0       1          0  ...   
1                 2      2020        5       23  30.0       1          0  ...   
2                 1      2019        5       29  28.0       1          0  ...   
3                 2      2020        4       22  49.0       1          0  ...   
4                 2      2020        1       10  41.0       1          0  ...   

   A  B  C  D  F  G  H  I  J  S  
0  1  0  0  0  0  0  0  0  0  0  
1  0  0  0  1  0  0  0  0  0  0  
2  0  1  0  0  0  0  0  0  0  0  
3  1  0  0  0  0  0  0  0  0  0  
4  0  1  0  0  0  0  0  0  0  0  

[5 rows x 48 columns]

In [77]:
df_HM_CustTrxnArtcls.dtypes

customer_id                    object
article_id                      int64
price                         float64
sales_channel_id                int64
txn_year                        int32
txn_mth                         int32
txn_day                         int32
age                           float64
ACTIVE                          uint8
LEFT CLUB                       uint8
PRE-CREATE                      uint8
Monthly                         uint8
NONE                            uint8
Regularly                       uint8
product_code                    int64
product_type_no                 int64
graphical_appearance_no         int64
colour_group_code               int64
perceived_colour_value_id       int64
perceived_colour_master_id      int64
department_no                   int64
index_group_no                  int64
section_no                      int64
garment_group_no                int64
Accessories                     uint8
Bags                            uint8
Cosmetic    

In [78]:
# Check for null values in the merged Trxn/Customer/Articles dataframe
[print('Any Null Values in the merged Trxn/Customer/Articles dataframe?','\n' ,df_HM_CustTrxnArtcls.isnull().any())]

Any Null Values in the merged Trxn/Customer/Articles dataframe? 
 customer_id                   False
article_id                    False
price                         False
sales_channel_id              False
txn_year                      False
txn_mth                       False
txn_day                       False
age                           False
ACTIVE                        False
LEFT CLUB                     False
PRE-CREATE                    False
Monthly                       False
NONE                          False
Regularly                     False
product_code                  False
product_type_no               False
graphical_appearance_no       False
colour_group_code             False
perceived_colour_value_id     False
perceived_colour_master_id    False
department_no                 False
index_group_no                False
section_no                    False
garment_group_no              False
Accessories                   False
Bags                          Fals

[None]

#### Final Data Preparation Steps

#### Remove Any Duplicates in Final Merged Dataframe. <a class="tocSkip">

In [79]:
# run a check to see if any duplicate entires exist in the merged Trxn/Customer/Articles dataframe
[print('Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe?','\n' 
                                                                       ,df_HM_CustTrxnArtcls.duplicated().sum())]

Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe? 
 4


[None]

In [80]:
# Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [81]:
duplicateRows

customer_id  article_id  \
17836  d00063b94dcb1342869d4994844a2742b5d62927f36843...   678342001   
17917  a1b5ea0495ee553afc1f2553c4db2c7a0ccc17da891c4c...   853391002   
18186  da5c4b0033207699fe4c59cefff41e4e5c18096592f3ac...   739144006   
30207  97a788e36864a5634c518ad1b1ddc87da50169cd11b625...   712587005   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
17836  0.006763                 1      2018       10       14  27.0       1   
17917  0.007983                 2      2020        6       29  28.0       1   
18186  0.010153                 2      2019       12        2  30.0       1   
30207  0.022017                 2      2019        7        1  27.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
17836          0  ...  0  0  0  1  0  0  0  0  0  0  
17917          0  ...  0  1  0  0  0  0  0  0  0  0  
18186          0  ...  1  0  0  0  0  0  0  0  0  0  
30207          0  ...  1  0  0  0  0  0  0  0  0  0  

[4 rows x 48 columns]

In [82]:
# Drop duplicate rows in the merged Trxn/Customer/Articles dataframe
df_HM_CustTrxnArtcls = df_HM_CustTrxnArtcls.drop_duplicates()

In [83]:
# Re-check: Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [84]:
duplicateRows

Empty DataFrame
Columns: [customer_id, article_id, price, sales_channel_id, txn_year, txn_mth, txn_day, age, ACTIVE, LEFT CLUB, PRE-CREATE, Monthly, NONE, Regularly, product_code, product_type_no, graphical_appearance_no, colour_group_code, perceived_colour_value_id, perceived_colour_master_id, department_no, index_group_no, section_no, garment_group_no, Accessories, Bags, Cosmetic, Furniture, Garment Full body, Garment Lower body, Garment Upper body, Items, Nightwear, Shoes, Socks & Tights, Swimwear, Underwear, Unknown, A, B, C, D, F, G, H, I, J, S]
Index: []

[0 rows x 48 columns]

In [85]:
# Confirm dimensions of Trxn/Customer/Articles (sample) dataframe AFTER removing duplicate rows
[print('\nCurrent Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed','\n' ,
                                                                                           df_HM_CustTrxnArtcls.shape)]


Current Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed 
 (31434, 48)


[None]

In [86]:
### See if there is a way to check the integrity of the customer merging..

###### 

## Split Data: Train and Test Sets

#### Split Data into 70% for Training, 30% for Testing. <a class="tocSkip">

In [87]:
# Split the data into train and test set
# Parameter set to 30% for size of Test set
# Given that there is ordering in the dataset, the 'shuffle' parameter is set to 'True'
train, test = train_test_split(df_HM_CustTrxnArtcls, test_size=0.3, random_state=101, shuffle=True)

In [88]:
# Output Confirmation of data split
dataDescription = 'H&M Sample Data Split'
print("\n{} Training Set Shape : \n\t".format(dataDescription))
print(train.shape)
print("\n{} Test Set Shape : \n\t".format(dataDescription))
print(test.shape)


H&M Sample Data Split Training Set Shape : 
	
(22003, 48)

H&M Sample Data Split Test Set Shape : 
	
(9431, 48)


###### 

## Model Preparation

#### Scale the Training Dataset (but not the Test set) <a class="tocSkip">

In [89]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [90]:
train.head()

customer_id  article_id  \
1029   e1992f521cd933be42cb0c2b07ed4c89f5de1f9b5ccd63...   592975011   
2360   cf3244bf58fd7cab321dda2d1e1a0955fb38d35268ca54...   659854007   
30562  4e628851f082849e5a7751e4d46bd65cbc9b2c3e931a5e...   610776022   
18240  069f79b30a8d8445ca0634b61404c6ded9473b22c336d5...   684209012   
27760  d65d5c414f64f65b45f4ad19fa16d974ec3e359d816dc2...   745475013   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
1029   0.050831                 2      2018        9       26  24.0       1   
2360   0.015237                 2      2019        4       29  46.0       1   
30562  0.008458                 1      2019       12       21  26.0       1   
18240  0.021339                 2      2019        4       10  21.0       1   
27760  0.033881                 1      2019        6       26  74.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
1029           0  ...  1  0  0  0  0  0  0  0  0  0  
2360           0  ...  0  1  0  0  0  0  0  0  0  0  
30562          0  ...  1  0  0  0  0  0  0  0  0  0  
18240          0  ...  0  1  0  0  0  0  0  0  0  0  
27760          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 48 columns]

In [91]:
chk_cols =  ['product_code',
            'product_type_no',
            'graphical_appearance_no',
            'colour_group_code',
            'perceived_colour_value_id',
            'perceived_colour_master_id',
            'department_no',
            'index_group_no',
            'section_no',
            'garment_group_no']

In [92]:
train[chk_cols]

product_code  product_type_no  graphical_appearance_no  \
1029         592975              272                  1010004   
2360         659854              298                  1010001   
30562        610776              255                  1010001   
18240        684209              298                  1010026   
27760        745475              265                  1010001   
...             ...              ...                      ...   
5695         734319              275                  1010001   
8006         635190               66                  1010016   
17745        697054              258                  1010016   
17933        613826              262                  1010016   
13151        519583              272                  1010023   

       colour_group_code  perceived_colour_value_id  \
1029                   6                          1   
2360                  12                          1   
30562                 12                          1   
18240                 10                          3   
27760                 73                          4   
...                  ...                        ...   
5695                  73                          4   
8006                   9                          4   
17745                 52                          2   
17933                 72                          2   
13151                 73                          2   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
1029                           12           1717               1          11   
2360                           11           4242               1          60   
30562                          11           1676               1          16   
18240                           9           4242               1          60   
27760                           2           1322               1          15   
...                           ...            ...             ...         ...   
5695                            2           1422               1          15   
8006                            5           3944               2          52   
17745                           4           1522               1          15   
17933                           2           1244               2          53   
13151                           2           1772               2          57   

       garment_group_no  
1029               1009  
2360               1018  
30562              1002  
18240              1018  
27760              1013  
...                 ...  
5695               1012  
8006               1019  
17745              1010  
17933              1007  
13151              1016  

[22003 rows x 10 columns]

In [93]:
cols_to_norm = ['price','age']

In [94]:
#train[cols_to_norm] = scaler.fit_transform(train[cols_to_norm])

In [95]:
# Normalizing numerical features so that each feature has mean 0 and variance 1
feature_scaler = StandardScaler()
#train[cols_to_norm] = feature_scaler.fit_transform(train[cols_to_norm])

In [96]:
#for x in dataset.columns[dataset.dtypes == 'int64']:
#    Dataset[x] = minmax.fit_transform(np.array(Dataset[I]).reshape(-1,1))

In [97]:
def scale_trainset_features(df, dtype, no_scale):
    for x in df.columns[df.dtypes == dtype]:
        if x != no_scale:
            df[x] = scaler.fit_transform(np.array(df[x]).reshape(-1,1))    

In [98]:
#for x in train.columns[train.dtypes == 'int64']:
#    if x != 'article_id':
#        train[x] = scaler.fit_transform(np.array(train[x]).reshape(-1,1))

In [99]:
no_scale = 'article_id'
scale_dtype = 'int64'
scale_trainset_features(train, scale_dtype, no_scale)

In [100]:
train.head()

customer_id  article_id  \
1029   e1992f521cd933be42cb0c2b07ed4c89f5de1f9b5ccd63...   592975011   
2360   cf3244bf58fd7cab321dda2d1e1a0955fb38d35268ca54...   659854007   
30562  4e628851f082849e5a7751e4d46bd65cbc9b2c3e931a5e...   610776022   
18240  069f79b30a8d8445ca0634b61404c6ded9473b22c336d5...   684209012   
27760  d65d5c414f64f65b45f4ad19fa16d974ec3e359d816dc2...   745475013   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
1029   0.050831               1.0      2018        9       26  24.0       1   
2360   0.015237               1.0      2019        4       29  46.0       1   
30562  0.008458               0.0      2019       12       21  26.0       1   
18240  0.021339               1.0      2019        4       10  21.0       1   
27760  0.033881               0.0      2019        6       26  74.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
1029           0  ...  1  0  0  0  0  0  0  0  0  0  
2360           0  ...  0  1  0  0  0  0  0  0  0  0  
30562          0  ...  1  0  0  0  0  0  0  0  0  0  
18240          0  ...  0  1  0  0  0  0  0  0  0  0  
27760          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 48 columns]

In [101]:
train[chk_cols]

product_code  product_type_no  graphical_appearance_no  \
1029       0.577823         0.512195                 0.999976   
2360       0.657633         0.560976                 0.999973   
30562      0.599066         0.480300                 0.999973   
18240      0.686698         0.560976                 0.999998   
27760      0.759810         0.499062                 0.999973   
...             ...              ...                      ...   
5695       0.746497         0.517824                 0.999973   
8006       0.628200         0.125704                 0.999988   
17745      0.702026         0.485929                 0.999988   
17933      0.602706         0.493433                 0.999988   
13151      0.490240         0.512195                 0.999995   

       colour_group_code  perceived_colour_value_id  \
1029            0.074468                      0.250   
2360            0.138298                      0.250   
30562           0.138298                      0.250   
18240           0.117021                      0.500   
27760           0.787234                      0.625   
...                  ...                        ...   
5695            0.787234                      0.625   
8006            0.106383                      0.625   
17745           0.563830                      0.375   
17933           0.776596                      0.375   
13151           0.787234                      0.375   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
1029                     0.619048       0.058716            0.00    0.094737   
2360                     0.571429       0.346040            0.00    0.610526   
30562                    0.571429       0.054051            0.00    0.147368   
18240                    0.476190       0.346040            0.00    0.610526   
27760                    0.142857       0.013769            0.00    0.136842   
...                           ...            ...             ...         ...   
5695                     0.142857       0.025148            0.00    0.136842   
8006                     0.285714       0.312130            0.04    0.526316   
17745                    0.238095       0.036527            0.00    0.136842   
17933                    0.142857       0.004893            0.04    0.536842   
13151                    0.142857       0.064975            0.04    0.578947   

       garment_group_no  
1029           0.333333  
2360           0.708333  
30562          0.041667  
18240          0.708333  
27760          0.500000  
...                 ...  
5695           0.458333  
8006           0.750000  
17745          0.375000  
17933          0.250000  
13151          0.625000  

[22003 rows x 10 columns]

In [102]:
train.head()

customer_id  article_id  \
1029   e1992f521cd933be42cb0c2b07ed4c89f5de1f9b5ccd63...   592975011   
2360   cf3244bf58fd7cab321dda2d1e1a0955fb38d35268ca54...   659854007   
30562  4e628851f082849e5a7751e4d46bd65cbc9b2c3e931a5e...   610776022   
18240  069f79b30a8d8445ca0634b61404c6ded9473b22c336d5...   684209012   
27760  d65d5c414f64f65b45f4ad19fa16d974ec3e359d816dc2...   745475013   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
1029   0.050831               1.0      2018        9       26  24.0       1   
2360   0.015237               1.0      2019        4       29  46.0       1   
30562  0.008458               0.0      2019       12       21  26.0       1   
18240  0.021339               1.0      2019        4       10  21.0       1   
27760  0.033881               0.0      2019        6       26  74.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
1029           0  ...  1  0  0  0  0  0  0  0  0  0  
2360           0  ...  0  1  0  0  0  0  0  0  0  0  
30562          0  ...  1  0  0  0  0  0  0  0  0  0  
18240          0  ...  0  1  0  0  0  0  0  0  0  0  
27760          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 48 columns]

In [103]:
#int32
#age                           float64

In [104]:
scale_dtype = 'int32'
scale_trainset_features(train, scale_dtype, no_scale)

In [105]:
scale_dtype = 'float64'
scale_trainset_features(train, scale_dtype, no_scale)

In [106]:
train.head() 

customer_id  article_id  \
1029   e1992f521cd933be42cb0c2b07ed4c89f5de1f9b5ccd63...   592975011   
2360   cf3244bf58fd7cab321dda2d1e1a0955fb38d35268ca54...   659854007   
30562  4e628851f082849e5a7751e4d46bd65cbc9b2c3e931a5e...   610776022   
18240  069f79b30a8d8445ca0634b61404c6ded9473b22c336d5...   684209012   
27760  d65d5c414f64f65b45f4ad19fa16d974ec3e359d816dc2...   745475013   

          price  sales_channel_id  txn_year   txn_mth   txn_day       age  \
1029   0.149592               1.0       0.0  0.727273  0.833333  0.115942   
2360   0.043927               1.0       0.5  0.272727  0.933333  0.434783   
30562  0.023800               0.0       0.5  1.000000  0.666667  0.144928   
18240  0.062041               1.0       0.5  0.272727  0.300000  0.072464   
27760  0.099275               0.0       0.5  0.454545  0.833333  0.840580   

       ACTIVE  LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
1029        1          0  ...  1  0  0  0  0  0  0  0  0  0  
2360        1          0  ...  0  1  0  0  0  0  0  0  0  0  
30562       1          0  ...  1  0  0  0  0  0  0  0  0  0  
18240       1          0  ...  0  1  0  0  0  0  0  0  0  0  
27760       1          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 48 columns]

###### 

###### 

###### 

###### 

## K Nearest Neighbor Recommendations

#### Set up function to find Nearest Neighbours (+ Articles bought) for a Customer <a class="tocSkip">

In [289]:
# Use a function to return a list of Nearest Neighbours for a given customer.
# Return Neighbour Customer Ids and the articles they both
# Set default parameters for NN algorithm - these can be overwritten during optimisationphase
def get_neighbourIDs_articles(query_index, NN_Neighbours, NN_metric = 'cosine', NN_algo = 'brute'):
     
    # Temp view of training set - for validation
    #train
    
    # Reset Training dataframe index - remove old one; this allows for extraction of correct index records
    # returned by the NearestNeighbors function
    train.reset_index(drop=True,inplace=True)
    
    # Temp view of training set - for validation
    #train
    
    # Set up numerical data inputs for NearestNeighbors function
    df_cust_articles = train[chk_cols]
    #df_cust_articles
    
    # Temp check on values - for verification
    train[['customer_id',
      'article_id',
      'product_code',
      'product_type_no',
      'graphical_appearance_no',
      'colour_group_code',
      'perceived_colour_value_id',
      'perceived_colour_master_id',
      'department_no',
      'index_group_no',
      'section_no',
      'garment_group_no']]
    #train[chk_cols]
    
    # Set up parameters of NearestNeighbors function
    #model_knn = NearestNeighbors(metric = 'cosine', algorithm = 'brute')
    model_knn = NearestNeighbors(metric = NN_metric, algorithm = NN_algo)
    
    # Implement NearestNeighbors function on merfe Trxn/Customer/Article data elements
    model_knn.fit(df_cust_articles)
    
    # Return index values for nearest neighbours - determined by 'Num_Neighbour' parameter
    distances, indices = model_knn.kneighbors(df_cust_articles.iloc[query_index,:].values.reshape(1, -1), 
                                              n_neighbors = NN_Neighbours)
    
    # Temp selection of a chosen record for which to find Nearest Neighbours
    #df_cust_articles.shape[0]
    #df_cust_articles.iloc[query_index,:].values.reshape(1, -1)
    #df_cust_articles.head()
    
    print('Query Index..')
    print(query_index)
    
    print('Indices values... ')
    print(indices)
    
    print('\ndf_cust_articles.loc[query_index]\n')
    print(df_cust_articles.loc[query_index])
    
    print('\n\ndf_cust_articles.loc[query_index,[product_code,product_type_no,graphical_appearance_no]]\n')
    print(df_cust_articles.loc[query_index,['product_code','product_type_no','graphical_appearance_no']])
    
    print('\n\ntrain.loc[query_index,[product_code,product_type_no,graphical_appearance_no]]\n')
    print(train.loc[query_index,['product_code','product_type_no','graphical_appearance_no']])
    
    print('\n\ndf_cust_articles.index[query_index]\n')
    for i in range(0, len(distances.flatten())):
        if i == 0:
            print('\nRecommendations for {0}:\n'.format(df_cust_articles.index[query_index])) # Again - what does this do?
        else:
            # What does the flatten do?
            print('df_cust_articles: {0}: Index: {1} {2} with distance of {3}:'.
              format(i,  
                     df_cust_articles.index[indices.flatten()[i]], 
                     df_cust_articles.loc[indices.flatten()[i],['product_code']],
                     distances.flatten()[i]))
        
    print('\n\ntrain.index[query_index]\n')    
    for i in range(0, len(distances.flatten())):
        if i == 0:
            print('\nRecommendations for {0}:\n'.format(train.index[query_index]))
        else:
            print('train: {0}: Index: {1} {2} {3} {4}, with distance of {5}:'
              .format(i,  
                      train.index[indices.flatten()[i]],
                      train.loc[indices.flatten()[i],['customer_id']], 
                      train.loc[indices.flatten()[i],['article_id']], 
                      train.loc[indices.flatten()[i],['product_code']], 
                      distances.flatten()[i]))    
        
    customer_list = []
    article_list  = []


    # Using the Index values returned by the Nearest neighbour model
    # Generate a list of Customer Ids and Article Ids for these neighbours
    for i in range(0, len(distances.flatten())):
        if i == 0:
            print('\nRecommendations for {0}:\n'.format(train.index[query_index]))
        else:
            print('Indices[i]: {}'.format(indices.flatten()[i]))
        
            print('{0}: Index: {1} {2} {3} {4}, with distance of {5}:'
              .format(i,  
                      train.index[indices.flatten()[i]],
                      train.loc[indices.flatten()[i],['customer_id']], 
                      train.loc[indices.flatten()[i],['article_id']], 
                      train.loc[indices.flatten()[i],['product_code']], 
                      distances.flatten()[i]))
            customer_list.append(train.loc[indices.flatten()[i],'customer_id'])
            article_list.append(train.loc[indices.flatten()[i],'article_id'])
        
    df_neighbourhood = pd.DataFrame({'customer_id': customer_list, 'article_id': article_list})
    del customer_list, article_list
    
    return df_neighbourhood  

#### Find Top Selling Products in 'Neighbourhood' for a Customer <a class="tocSkip">

In [299]:
# Function to return Top 12 Articles Purchased by the Customer 'Neighbourhood'
def get_Top12_Articles_from_CustNNs(df, num_of_articles):
    
    print('get_Top12_Articles_from_CustNNs')
    
    # Finding number of times a particular item is purchased by customers in the neighbourhood
    temp = df.groupby(['article_id'])['customer_id'].agg('count').reset_index() 
    temp.columns = ['article_id','count']
    
    # Sort the articles by the number of time purchased by the 'neighbourhood' customers
    temp = temp.sort_values('count', ascending=False)
    
    # Temp Print out of Article Id List
    print(temp)
    
    # Select the Top 12 - if there are duplicate just select the first entry
    temp = temp.nlargest(num_of_articles,'count', keep='first')
    
    # Isolate the Article Ids
    temp = temp['article_id']
    
    # Convert to dataframe and reset index
    df_article_list = pd.DataFrame({'article_id': temp})
    df_article_list = df_article_list.reset_index(drop=True)
    
    
    # Return the list of Top 12 articles
    return df_article_list

In [314]:
# Call function to find nearest neighbour Customer Ids and Article Ids
# Pass in index, parameters for Nearest Neighbour function, and the number of neighbours to retrun
df_neighbours = get_neighbourIDs_articles(3, 100, 'cosine','brute')

Query Index..
3
Indices values... 
[[    3 16253 19034 11419  3025  5373  3771 14386 10409 19275 20641 14032
   5918 16249 21482 17277 11090 20412  7754  3409  7396 20120  4356  4212
   3064 15807 11014 13934  9731 10097 15395  2418 20902 16815  8185 12285
   2613 17324 15051 19248  4317 19428 13694  7999 13630  1443  4134 13527
   6178  8715  7406  8420  9390  6567  4184 13667 10547   165 16994  9404
  10212  7441 13603  8447 12807 21512 20445 17598  3745    10 17280 14279
   7717  6924 17561 20223  1829 20008 21670  8747 21242 16706 20739 11560
    818 14787 17060 15532  1680 11107 14018 13896  2542 20261  1580  4967
   7354 21541  9268   749]]

df_cust_articles.loc[query_index]

product_code                  0.686698
product_type_no               0.560976
graphical_appearance_no       0.999998
colour_group_code             0.117021
perceived_colour_value_id     0.500000
perceived_colour_master_id    0.476190
department_no                 0.346040
index_group_no                0.0000

train: 69: Index: 10 customer_id    1a4abe9ba0ee4dc613a46d42b4c8a0dc2834604fd81598...
Name: 10, dtype: object article_id    619285005
Name: 10, dtype: object product_code    0.60922
Name: 10, dtype: object, with distance of 0.001974533115644661:
train: 70: Index: 17280 customer_id    5ba433496ca1a84d217de4f382af9f1b0e2e0a492fa837...
Name: 17280, dtype: object article_id    757158001
Name: 17280, dtype: object product_code    0.773752
Name: 17280, dtype: object, with distance of 0.0020017149340142426:
train: 71: Index: 14279 customer_id    ef16d8241d8267626fd2637155dd57ec09f778423373b9...
Name: 14279, dtype: object article_id    757614002
Name: 14279, dtype: object product_code    0.774296
Name: 14279, dtype: object, with distance of 0.0020147868348125986:
train: 72: Index: 7717 customer_id    09b22b6b0c50462049c6d08445c00c813a1369c7880c7e...
Name: 7717, dtype: object article_id    757614002
Name: 7717, dtype: object product_code    0.774296
Name: 7717, dtype: object, with distance of 0

Indices[i]: 11090
16: Index: 11090 customer_id    67840996ccfedbd8d398e3861393510e1cd4848b8df978...
Name: 11090, dtype: object article_id    659854010
Name: 11090, dtype: object product_code    0.657633
Name: 11090, dtype: object, with distance of 0.00013149448951454978:
Indices[i]: 20412
17: Index: 20412 customer_id    fb6b1def88f628b132d546d68d89be63a1972983729838...
Name: 20412, dtype: object article_id    659854010
Name: 20412, dtype: object product_code    0.657633
Name: 20412, dtype: object, with distance of 0.00013149448951454978:
Indices[i]: 7754
18: Index: 7754 customer_id    c77c7325778b91c2e3414cffbf6940b4b537d6eceaed98...
Name: 7754, dtype: object article_id    653664008
Name: 7754, dtype: object product_code    0.650246
Name: 7754, dtype: object, with distance of 0.00017644864456689557:
Indices[i]: 3409
19: Index: 3409 customer_id    e699d3f41a83a50b2a21f302e3bdf05cccf796dd718fc3...
Name: 3409, dtype: object article_id    717370003
Name: 3409, dtype: object product_code   

47: Index: 13527 customer_id    df02770f8b5ba2441937338f8fe237c597d764a3771ca4...
Name: 13527, dtype: object article_id    776237029
Name: 13527, dtype: object product_code    0.79652
Name: 13527, dtype: object, with distance of 0.0015043361889733875:
Indices[i]: 6178
48: Index: 6178 customer_id    778e96586c2b38758650f02299d0a221f2915eeaa1d998...
Name: 6178, dtype: object article_id    776237027
Name: 6178, dtype: object product_code    0.79652
Name: 6178, dtype: object, with distance of 0.0015043361889733875:
Indices[i]: 8715
49: Index: 8715 customer_id    604a3e5bdc04d072e3bf4ad9a02e2673ba18dd843a4f84...
Name: 8715, dtype: object article_id    780887001
Name: 8715, dtype: object product_code    0.802069
Name: 8715, dtype: object, with distance of 0.001655919385337823:
Indices[i]: 7406
50: Index: 7406 customer_id    b636c4324acc55d4f7bae33017f2c6b0da06235acb2ea3...
Name: 7406, dtype: object article_id    653188002
Name: 7406, dtype: object product_code    0.649678
Name: 7406, dtype: 

In [315]:
#Temp Check 
df_neighbours

customer_id  article_id
0   0602fd1795c17f410689895a985781fcdddef71edd4b7a...   684209012
1   8ccce9d55e7dca04179749d3cbeb56b3a178c70f2a3448...   684209020
2   833b0fcd13fa92088e40e8ca8f53bd9e8259e0b03a9768...   698286007
3   4b9cfa1ed7244336f3ea2d6ca33246a27bbe3a65222fc3...   698286007
4   b6fa7a99497c2d79e745d8ecc87eda2c788a57329e3497...   698286007
..                                                ...         ...
94  bba138319aed9c807ad9ff94deeebda0f6bb3aa233e198...   559630004
95  5c8e4398cbeafafa14674dd55e9eaa22b084cda26a503b...   559630020
96  c525e56bd1620a431d620fffd44a47ccd23cd697e842a7...   559630020
97  c6733051229dde65d23924b3e88d5dd4e7080d5acd0b43...   559616006
98  f2cc8720d5508fea5e5cc664a6040878c0ae1bffdb3bf7...   559616006

[99 rows x 2 columns]

In [316]:
#df_neighbours.dtypes

In [317]:
# Call function to return list of the most popular Articles in terms of the 
# purchases made by the customers neighbourhood
df_cust_preds = get_Top12_Articles_from_CustNNs(df_neighbours, 12)

get_Top12_Articles_from_CustNNs
    article_id  count
21   659854010      8
23   678942016      6
31   698286007      3
19   653188002      3
52   774785002      3
..         ...    ...
35   717370003      1
9    589017003      1
37   723084004      1
8    585480003      1
62   812668001      1

[63 rows x 2 columns]


In [318]:
df_cust_preds

article_id
0    659854010
1    678942016
2    698286007
3    653188002
4    774785002
5    744786002
6    728473001
7    589599002
8    811925005
9    811835006
10   758118001
11   757614002

In [296]:
#df_cust_preds.dtypes

article_id    int64
dtype: object

###### 

###### 

###### 

## Submission

#### The solution file to have 1371980 prediction rows - use Sample Submission to construct  <a class="tocSkip">

In [132]:
# Read in Sample Submission File
df_sub  = pd.read_csv('../input/sample_submission.csv',
                            #usecols= ['customer_id'], 
                            dtype={'customer_id': 'string'})

In [133]:
numCustomers = df_sub.shape[0]

In [134]:
# Temp code to build submission framework
df_sub_sample = df_sub.sample(frac=0.001, replace=False)
uniBin = 1
listUniBins = [1]

In [135]:
# Set up folder for submission file with predictions
submission_path = '../submission'

In [136]:
# Create folder above project directory
os.makedirs(submission_path, exist_ok=True)  

In [137]:
# Wrtite submission file to disk
df_sub_sample.to_csv(submission_path+'/submission_' + str(uniBin) + '.csv',index=False)

In [138]:
# Check submission file size is correct
for i, uniBin in enumerate(listUniBins):
    dfTemp  = pd.read_csv(submission_path+'/submission_' + str(uniBin) + '.csv')
    if i == 0:
        dfSub = dfTemp
    else:
        dfSub = cudf.concat([dfSub, dfTemp], axis=0)

[print("Matching the number of dfSub rows to the correct submission size. \nSubmission Row: {} v Expected Rows: {}"
                                                                                   .format(dfSub.shape[0],numCustomers))]

Matching the number of dfSub rows to the correct submission size. 
Submission Row: 1372 v Expected Rows: 1371980


[None]

In [139]:
# Check format of submission file
dfCheck = pd.read_csv(submission_path+'/submission_' + str(uniBin) + '.csv')
dfCheck.head()

customer_id  \
0  435e2470814dea9a65afaa7541a5c133cbb8992390af73...   
1  da567ef9df7631ef5b1030c17a76208936e002fb9b4ec2...   
2  bf88621e99ba92eb03d5844694cb438ce222856019af99...   
3  090eaf056fea329e80737fb24c25a9a3bd0b4efcc2a763...   
4  16f15e392bbf8e2eafc554351a7d08cf757f4043c6526e...   

                                          prediction  
0  0706016001 0706016002 0372860001 0610776002 07...  
1  0706016001 0706016002 0372860001 0610776002 07...  
2  0706016001 0706016002 0372860001 0610776002 07...  
3  0706016001 0706016002 0372860001 0610776002 07...  
4  0706016001 0706016002 0372860001 0610776002 07...

###### 

###### 

## References

#### <a class="tocSkip">